### ISSUE TITLE
Consider refactoring how functions are passed to azure-webjobs-sdk

### ISSUE THREAD

**Issue by tpetricek** [Azure/azure-webjobs-sdk-script#332]
As part of working on #279, I was trying to understand how the functions are invoked when an event happens. My current understanding is that the code generates a type using reflection emit that is then analyzed somewhere else and turned into an index. This seems very complicated. My understanding is:

 * In `WebJobs.Script`, the [Generate method in FunctionGenerator](https://github.com/Azure/azure-webjobs-sdk-script/blob/ea27cc0e32e5d883d6dc1d40430a6e9add85beea/src/WebJobs.Script/Description/FunctionGenerator.cs#L27) uses Reflection Emit to produce a type with a static method for every Azure function.

 * This type is then passed to the runtime in [azure-webjobs-sdk](https://github.com/Azure/azure-webjobs-sdk) via an [ITypeLocator instance](https://github.com/Azure/azure-webjobs-sdk-script/blob/ea27cc0e32e5d883d6dc1d40430a6e9add85beea/src/WebJobs.Script/Config/TypeLocator.cs).

 * In azure-webjobs-sdk, the type locator is then passed to [FunctionIndexProvoider](https://github.com/Azure/azure-webjobs-sdk/blob/b1fb20b0edf0530d62127d6dd9f50cc9345ca7b6/src/Microsoft.Azure.WebJobs.Host/Indexers/FunctionIndexProvider.cs#L15), which gets all the types & their methods and eventually invokes `IndexMethodAsyncCore` [here](https://github.com/Azure/azure-webjobs-sdk/blob/b1fb20b0edf0530d62127d6dd9f50cc9345ca7b6/src/Microsoft.Azure.WebJobs.Host/Indexers/FunctionIndexer.cs#L164).

 * This uses reflection (over the type that we had to manually generate with reflection emit!) to find the parameters of the method and figure out things like triggers for the function. At the end, it just makes `IFunctionDefinition` and `FunctionDescriptor`.

I do not know all the background here, but this seems rather complicated, especially given the fact that `FunctionDescriptor` is pretty simple type (with just list of parameters and `MethodInfo`).

In a perfect world, I think that azure-webjobs-sdk-script should not need to use reflection emit at all. It could just return an instance of some simple interface that defines what the parameters are and has an invoke method that gets called when the function is triggered. Using `System.Type` and `MethodInfo` as the public protocol makes the code quite hard to follow & debug. 

I'm not immediately planning to do this, because I think it's a big change for a first-time contributor, but I think it would make the life of everyone who wants to contribute much easier.

**Comment by mathewc** [comment]
Yes, this is something we've discussed. The current meta model for describing job functions in the core WebJobs SDK is reflection. I've considered refactoring everything to allow for a pluggable description model, which would support description via JSON metadata.

That is a substantial refactoring of the core SDK however, so the IL emit model was the most direct way of moving forward. In the fullness of time we might get to the SDK refactor, but there are much higher priority things for us to work on, given that the existing mechanism isn't causing any problems.

**Comment by mathewc** [comment]
Also, I think this issue should be moved to the core SDK repo, since this is really a feature suggestion for that repo.

**Comment by christopheranderson** [comment]
Moved to the SDK repo to better track this work - https://github.com/Azure/azure-webjobs-sdk/issues/714
